### ISSUE TITLE
400% CPU load on "nexus5" "d4", strings with numbers in them.

### ISSUE THREAD

**Issue by carstenhag** [lidel/meowkov#8]
Hey, basically I'm trying out this bot right now and it's fun!

Just that on some specific strings it goes wham and uses 4x100% cpu time: https://i.imgur.com/NW5pLAF.png

Any idea what could be causing this? Corpus is 43MB of irc logs. Meowkov is running in a VMware vm on an overclocked i5 PC.

The string that causes this is "nexus5". "nexus 5" does not invoke this.

Screen recording: https://imgrush.com/tpH7OV5UCfBf.mp4

After ~30 seconds it answered "<GoBot> Moter8: " :dash: :D

**Comment by carstenhag** [comment]
Also, I tried adding "nexus5" to the corpus and tried to reproduce and it happened again :/

**Comment by lidel** [comment]
When meowkov is unable to produce at least `MinResponsePool` responses to choose from via regular Markov chain logic, an additional heuristic kicks-in: `seeds = artificialSeed(input, power)`. This always happens when your input is just one word (shorter than `ChainLength`), or you used rare word that is not in many corpus keys (key is built from  `ChainLength`of words, so you need something more than a single-word input).

The way artificial seeding works is this: it picks `power` random keys from corpus, and generates additional seeds by appending those keys to the initial input. Then uses it as an input for regular Markov chain logic. It is repeated until enough results is produced or `MaxResponseTries` limit is hit.

This usually gives some satisfying results in first two or three iterations, but if your input is very short (eg. only one word that is not in corpus) and rare, it may take more time. With each iteration meowkov rapidly increases the amount of random keys (`power`), which impacts CPU usage significantly. It is very brutal but effective technique as _it does its best to produce response that is somehow linked to input_ based on historical data (corpus).  

To summarise: high CPU is what we expect during `artificialSeed` process (meowkov uses all CPU cores for performance). Personally I don't mind if it takes a moment to produce response in such edge-cases, as it gives an impression of bot 'thinking' and better results on average, but you can disable it via `meowkow.conf`  (see below).

**TL;DR:** If you want to save CPU usage during such edge-cases, simply decrease the amount of retry iterations (`MaxResponseTries`, set to 1 to disable retries) and response pool size  (`MinResponsePool`).

Setting `MaxResponseTries=0` and `MinResponsePool=1` will disable this CPU-hungry heuristics and run only basic Markov chain logic.  Be aware that depending on your corpus, it will impact the quality of produced gibberish :-)

Let me know if it helped.

**Comment by carstenhag** [comment]
Ah I see, thanks for the very extensive answer. But, shouldn't it return something after all that computation? Or well, if it can't compute something, give out some error or "something" -- just not an empty message.

**Comment by lidel** [comment]
Result depends on the corpus:  sometimes it is just a bad luck (Markov chains are statistical tool) and even with multiple iterations and artificial seeding there are no replies.
If meowkov runs out of retries and still does not have any good response it should return `randomSmiley()`. (If it does not, then we have a bug :-))

**Comment by carstenhag** [comment]
Oohhhh! Sorry, I just remembered that I removed those smileys and just let a "" in there, haha. Thanks.
